### PR TITLE
Propagate `collate_fn` to `InputTransform` in `ImageEmbedder`

### PR DESCRIPTION
**Pull request by krshrimali**
## What does this PR do?

Looks like `collate_fn` was never propagated before. The reason it worked before: `pretraining_transform` was never called, hence the inputs were always tensors and not a list of tensors (so no need to collate).

Fixes https://github.com/PyTorchLightning/lightning-flash/issues/1218.

## Before submitting
- [x] Was this **discussed/approved** via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the **[contributor guideline](https://github.com/PyTorchLightning/lightning-flash/tree/master/.github/CONTRIBUTING.md)**, Pull Request section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes?
- [x] Did you write any **new necessary tests**? [not needed for typos/docs]
- [x] Did you verify **new and existing tests pass** locally with your changes?
- [x] If you made a notable change (that affects users), did you **update the [CHANGELOG](https://github.com/PyTorchLightning/lightning-flash/blob/master/CHANGELOG.md)**?

## PR review
 - [x] Is this pull request **ready for review**? (if not, please submit in draft mode)

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: flash/image/embedding/losses/vissl_losses.py [hunk 1]
```diff
@@ -27,7 +26,6 @@
 
 
 def get_loss_fn(loss_name: str, cfg: AttrDict):
-    set_cpu_device()
     loss_fn = LOSS_REGISTRY[loss_name](cfg)
```
Comment (krshrimali) [inline flash/image/embedding/losses/vissl_losses.py:29]: The `__init__` function of SimCLR loss criterion pre-computes some masks, and setting CPU device explicitly here led us to get devices mismatch error for further calculates (masks were on CPU, while other computations were on CUDA).

File: flash/image/embedding/model.py [hunk 2]
```diff
@@ -112,9 +114,13 @@ def __init__(
...
 
-        input_transform, self.collate_fn = self.transforms.get(pretraining_transform)(**pretraining_transform_kwargs)
+        input_transform, self.transform_collate_fn = self.transforms.get(pretraining_transform)(
+            **pretraining_transform_kwargs
+        )
         output = ApplyToKeys(DataKeys.INPUT, input_transform)
-        self.input_transform = partial(LambdaInputTransform, transform=output)
+        self.input_transform = partial(
+            LambdaInputTransform, transform_collate_fn=self.transform_collate_fn, transform=output
+        )
```
Comment (krshrimali) [inline flash/image/embedding/model.py:123]: We were not using the collate function before...

File: flash/image/embedding/vissl/adapter.py [hunk 3]
```diff
@@ -122,13 +122,6 @@ def from_task(
 
         return result
 
-    def on_epoch_start(self) -> None:
-        use_gpu = self.adapter_task.device != torch.device("cpu") and self.adapter_task.device != "cpu"
-        if hasattr(self.loss_fn, "info_criterion"):
-            self.loss_fn.info_criterion.use_gpu = use_gpu
-        if hasattr(self.loss_fn, "swav_criterion"):
-            self.loss_fn.swav_criterion.use_gpu = use_gpu
-
```
Comment (krshrimali) [inline flash/image/embedding/vissl/adapter.py:131]: I assume we don't need this anymore? Waiting for CI...

File: flash/image/embedding/vissl/transforms/utilities.py [hunk 4]
```diff
@@ -46,7 +46,9 @@ def multicrop_collate_fn(samples):
     for idx, ele in enumerate(inputs):
         inputs[idx] = torch.stack(ele)
 
-    result[DataKeys.INPUT] = inputs
+    # TODO @krshrimali: This can be slow
+    for input in inputs:
+        result[DataKeys.INPUT] = torch.as_tensor(input, device=input.device, dtype=input.dtype)
```
Comment (krshrimali) [inline flash/image/embedding/vissl/transforms/utilities.py:51]: This will be slow if `inputs` is huge!

File: tests/image/embedding/test_model.py [hunk 5]
```diff
@@ -89,3 +89,46 @@ def test_vissl_training(backbone, training_strategy, head, pretraining_transform
...
+        head=head,
+        pretraining_transform=pretraining_transform,
+        training_strategy_kwargs=training_strategy_kwargs,
+    )
+
+    trainer = flash.Trainer(max_steps=3, max_epochs=1, gpus=torch.cuda.device_count())
+
+    predict_datamodule = ImageClassificationData.from_datasets(predict_dataset=FakeData(), batch_size=4)
+    trainer.fit(embedder, datamodule=datamodule)
+    trainer.predict(embedder, datamodule=predict_datamodule)
```
Comment (krshrimali) [inline tests/image/embedding/test_model.py:134]: Could have done this in the test that existed before but thought decoupling training test vs prediction test is a good idea?

**Review by ethanwharris — APPROVED** [review]
Awesome, LGTM :smiley:
